src/transformStateWithClones.js

```diff
@@ -1,13 +1,33 @@
 'use strict';
 
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    for (const key in action) {
+      if (action[key] === 'addProperties') {
+        Object.assign(newState, action.extraData);
+        stateHistory.push(structuredClone(newState));
+        break;
+      }
+
+      if (action[key] === 'removeProperties') {
+        for (const keyToRemove of action.keysToRemove) {
```

Comment:
Ensure that `action.keysToRemove` is an array. If there's a possibility that it might not be, you should add a check or validation to handle such cases.